tools/add_triage_signature.py

```diff
@@ -383,8 +384,87 @@ class StorageDetailSignature(Signature):
+            if cls.SmartctlExitCode(exit_code) == cls.SmartctlExitCode.SOME_ATA_COMMAND_FAILED_OR_SMART_CHECKSUM_ERROR:
+                if "smart_status" not in smart:
+                    output.append(
+                        "{color:green}*this typically happens on virtual disks*{color} that don't actually have S.M.A.R.T. information,"
+                        " but may also happen due to other reasons")
+
+        if "messages" in smartctl:
+            for message in smartctl["messages"]:
+                severity = message.get("severity", "unknown")
+                message_string = message.get("string", "")
+                output.append(f'smartctl message with severity "{severity}": "{message_string}"')
+
+        if "smart_status" in smart:
+            passed = smart["smart_status"]["passed"]
+            if passed:
+                if "model_name" in smart:
+                    if smart["model_name"] == "QEMU HARDDISK":
+                        return "{color:green}*Virtual QEMU HARDDISK pretending to have S.M.A.R.T.*{color}"
+
+                output.append("{color:green}*Disk S.M.A.R.T. status OK*{color}")
+            else:
+                output.append("{color:red}*Disk S.M.A.R.T. status BAD*{color}")
+
+        if "power_cycle_count" in smart:
+            output.append(f"disk has been power cycled {smart['power_cycle_count']} times throughout its life")
+
+        if "power_on_time" in smart:
+            if "hours" in smart["power_on_time"]:
+                hours = smart["power_on_time"]["hours"]
+                output.append(
+                    f"disk has been powered on for {hours // 24} days and {hours % 24} hours throughout its life")
+
+        if "ata_smart_attributes" in smart:
+            table = smart["ata_smart_attributes"].get("table", [])
+            interesting = ("Program_Fail_Count", "Erase_Fail_Count", "Offline_Uncorrectable")
+            table_filtered = [entry for entry in table if
+                              entry.get("name") in interesting and
+                              entry.get("raw")["value"] > 0]
 
+            output.extend(f'ata_smart_attribute *{entry["name"]}* has notable value *{entry["raw"]["value"]}*'
+                          for entry in table_filtered)
+
+        if "nvme_smart_health_information_log" in smart:
+            nvme_log = smart["nvme_smart_health_information_log"]
+            if "percentage_used" in nvme_log:
+                output.append(f"SSD percentage used: %{nvme_log['percentage_used']}")
+
+        return ", ".join(output)
+
+    @classmethod
+    def _parse_smart(cls, smart_json):
+        try:
+            return cls._parse_smart_internal(json.loads(smart_json))
+        except json.JSONDecodeError as e:
+            return f"Failed to decode S.M.A.R.T. JSON, error: {e}"
+        except Exception as e:
+            return f"S.M.A.R.T. JSON has unexpected format, error: {e}"
```

Comment:
above exception should be JSON decode error, no?

Reply:
No, the second message I meant that the structure of the JSON is unexpected, not that the JSON encoding itself is wrong.

Reply:
Improved, somewhat